dd-java-agent/instrumentation/jetty-6/src/main/java/datadog/trace/instrumentation/jetty6/JettyHandlerAdvice.java

```diff
@@ -0,0 +1,95 @@
+// Modified by SignalFx
+package datadog.trace.instrumentation.jetty6;
+
+import static datadog.trace.instrumentation.jetty6.JettyDecorator.DECORATE;
+
+import datadog.trace.api.DDTags;
+import datadog.trace.context.TraceScope;
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.propagation.Format;
+import io.opentracing.tag.Tags;
+import io.opentracing.util.GlobalTracer;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import net.bytebuddy.asm.Advice;
+import org.mortbay.jetty.HttpConnection;
+import org.mortbay.jetty.Response;
+
+public class JettyHandlerAdvice {
+
+  @Advice.OnMethodEnter(suppress = Throwable.class)
+  public static Scope startSpan(
+      @Advice.This final Object source, @Advice.Argument(1) final HttpServletRequest req) {
+
+    System.out.println("Starting 6 span");
+
+    if (GlobalTracer.get().activeSpan() != null) {
+      // Tracing might already be applied.  If so ignore this.
+      return null;
+    }
+
+    final SpanContext extractedContext =
+        GlobalTracer.get()
+            .extract(Format.Builtin.HTTP_HEADERS, new HttpServletRequestExtractAdapter(req));
+    final Scope scope =
+        GlobalTracer.get()
+            .buildSpan("jetty.request")
+            .asChildOf(extractedContext)
+            .withTag("span.origin.type", source.getClass().getName())
+            .startActive(false);
+
+    final Span span = scope.span();
+    final String resourceName = req.getMethod() + " " + source.getClass().getName();
+    span.setTag(DDTags.RESOURCE_NAME, resourceName);
+
+    DECORATE.afterStart(span);
+    DECORATE.onConnection(span, req);
+    // http url must be set after resource name for URLAsResourceName decorator
+    DECORATE.onRequest(span, req);
+
+    if (scope instanceof TraceScope) {
+      ((TraceScope) scope).setAsyncPropagation(true);
+    }
+    return scope;
+  }
+
+  @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+  public static void stopSpan(
+      @Advice.Argument(1) final HttpServletRequest req,
+      @Advice.Argument(2) final HttpServletResponse resp,
+      @Advice.Enter final Scope scope,
+      @Advice.Thrown final Throwable throwable) {
+
+    if (scope != null) {
+      final Span span = scope.span();
+
+      if (req.getUserPrincipal() != null) {
+        span.setTag(DDTags.USER_NAME, req.getUserPrincipal().getName());
+      }
+
+      Response response = HttpConnection.getCurrentConnection().getResponse();
+      DECORATE.onResponse(span, resp);
+
+      if (throwable != null) {
+        if (response.getStatus() == HttpServletResponse.SC_OK) {
+          // exception is thrown in filter chain, but status code is incorrect
+          Tags.HTTP_STATUS.set(span, 500);
```

Comment:
Probably best to only set the status if we're certain it's what's in the response.

Reply:
Removed this, but that allows weird looking situations where error tags are present with a 200 status.

Reply:
I think removing `DECORATE.onResponse()` would prevent setting the status code unless it's known to be valid.  Is async support necessary to defer finishing until the response code is correct?

Reply:
Moved this to only be called if no exception was thrown